### PR TITLE
Remove empty list as default, add support for tuples

### PR DESCRIPTION
**Pull request by lhenault**
Should : 
- address #23
- add support for tuples
- Switch to `NotImplementedError` instead of random things in default model classes

@Nintorac lmk what you think :)

### CODE REVIEW

File: src/simple_ai/models.py [hunk 1]
```diff
@@ -113,7 +114,7 @@ class RpcChatLanguageModel:
 
     def chat(
         self,
-        messages: list[list[str]] = [],
+        messages: TupleOrList[TupleOrList[str]] = (),
```
Comment (Nintorac) [inline src/simple_ai/models.py:117]: Off the top of my head I believe this is a list of dicts, does it need a default?

File: src/simple_ai/models.py [hunk 2]
```diff
@@ -140,7 +141,7 @@ def chat(
 
     def stream(
         self,
-        messages: list[list[str]] = [],
+        messages: TupleOrList[TupleOrList[str]] = (),
```
Comment (Nintorac) [inline src/simple_ai/models.py:144]: same a line 117

> Off the top of my head I believe this is a list of dicts, does it need a default?

File: src/simple_ai/api/grpc/chat/model.py [hunk 3]
```diff
@@ -1,13 +1,12 @@
...
-            {"role": message.get("role"), "content": message.get("content")[::-1]}
-            for message in chatlog
-        ]
-
-    def stream(self, chatlog: list[list[str]] = [], *args, **kwargs) -> list:
-        raise NotImplementedError()
+    def chat(self, chatlog: TupleOrList[TupleOrList[str]] = (), *args, **kwargs) -> list:
+        raise NotImplementedError
+
+    def stream(self, chatlog: TupleOrList[TupleOrList[str]] = (), *args, **kwargs) -> list:
```
Comment (Nintorac) [inline src/simple_ai/api/grpc/chat/model.py:11]: see comment  line 117 of `src/simple_ai/models.py`

File: src/simple_ai/api/grpc/chat/client.py [hunk 4]
```diff
@@ -21,7 +22,7 @@ def get_chatlog(stub, chatlog):
 
 def run(
     url: str = "localhost:50051",
-    messages: List[List[str]] = [],
+    messages: TupleOrList[TupleOrList[str]] = (),
```
Comment (Nintorac) [inline src/simple_ai/api/grpc/chat/client.py:25]: see comment line 117 of `src/simple_ai/models.py`

File: src/simple_ai/api/grpc/chat/client.py [hunk 5]
```diff
@@ -62,7 +63,7 @@ def stream_chatlog(stub, chatlog):
 
 def run_stream(
     url: str = "localhost:50051",
-    messages: List[List[str]] = [],
+    messages: TupleOrList[TupleOrList[str]] = (),
```
Comment (Nintorac) [inline src/simple_ai/api/grpc/chat/client.py:66]: see comment line 117 of `src/simple_ai/models.py`

**Review by Nintorac — COMMENTED** [review]
What's the motivation to use tuple as the default value rather than optional? I have not seen that pattern before

**Comment by lhenault** [comment]
> What's the motivation to use tuple as the default value rather than optional? I have not seen that pattern before

Mostly "two birds one stone": adding the support for `tuple` on top of `list` (I mean why not?) *and* no more having a mutable default. Agree that I haven't seen this pattern before myself, but didn't see any issue with it.

File: src/simple_ai/models.py (hunk 1 above)
Comment (lhenault) [reply src/simple_ai/models.py:117]: I believe it's a list of lists, not dicts, on the gRPC end, but anyway might actually be just `Optional`.

**Comment by Nintorac** [comment]
hmm, would err on the side of keep it simple. Will there ever be a tuple passed in or is that only representing `None`? It doesn't matter too much though so will leave it up to you :)

**Pull request closed without merging**